utils/measure/runner/charging.py

```diff
@@ -43,8 +43,29 @@ def run(
         )
         input("Hit enter when you are ready to start..")
 
+        print()
+
         battery_level = self.controller.get_battery_level()
         measurements: dict[int, list[float]] = {}
+        is_charging = self.controller.is_charging()
+        is_valid_state = self.controller.is_valid_state()
+        wait_message_printed = False
+
+        if battery_level < 100:
+            while not is_charging:
+                if not is_valid_state:
+                    raise RunnerError("Device is not in a valid state.")
+
+                if not wait_message_printed:
+                    print("waiting for vacuum cleaner to start charging...")
+                    wait_message_printed = True
+
+                time.sleep(1)
+                is_charging = self.controller.is_charging()
+                is_valid_state = self.controller.is_valid_state()
```

Comment:
```suggestion
        wait_message_printed = False

        if battery_level < 100:
            while not is_charging:
                if not self.controller.is_valid_state():
                    raise RunnerError("Device is not in a valid state.")

                if not wait_message_printed:
                    print("waiting for vacuum cleaner to start charging...")
                    wait_message_printed = True

                time.sleep(1)
                is_charging = self.controller.is_charging()
```

We can evaluate `is_valid_state` directly on one place, as it's also only called at line 56. No need to assign to variable.